### ISSUE TITLE
not working

### ISSUE THREAD

**Issue by Renjie8823** [edwinpalegre/EE8204-ResUNet#1]
Hi, when I tried to run the patch_train_val.py. The line 46 next(os.walk(os.path.join(train_dataset, 'image'))) went wrong. It said that stop iteration. Do you know how to solve it ?
Could you explain in detail how to generate 30,000 samples? I don't quite understand your code

**Comment by edwinpalegre** [comment]
Hi Renjie,

Have you downloaded the image and mask datasets? If so, have you set up your directories based on the code's folder names? If so, rerun the code again. I have encountered StopIteration error before when initially running the code but running it a second time immediately after the StopIteration error seems to overcome this issue. I have been meaning to look at this but haven't had the time. I have a feeling it's caused by the code not updating the path variables properly until after the code executes. If it still doesn't work let me know, I did just run it on my end and confirmed it works so there might be something else happening.

In terms of how it runs, first and foremost we point the program to the proper path that houses hte original Mass. Roads dataset. These images are 1500x1500 which are far to large to actually use with my current GPU. Hence, we need to generate sampled patches of them that are smaller. Lines 34-40 just check if the directories that will house the sampled images & mask exist, if they don't Python makes it. Next we iterate through the original images in line 46. This creates a list with all the names of the images. Note that the images and masks are named 1.png, 2.png, 3.png, etc since we will be calling them based on number. The image and its corresponding mask share the same name. It makes it much easier to handle this way.

I made a function called get_randompatch(). It first randomly chooses an image, since our image names are numbers we can set the range (1, number of images + 1). We then update the local path which points to the image and the corresponding mask folders. We open up the image and its corresponding mask together and store their sizes. Then we set the coordinates it needs to use to crop the image. By subtracting the desired image size (in this case 512 or 224) from the original image size, we are limiting the start coordinate from choosing a value that would cause the sampling to go outside the image. We multiply the start value with np.random.rand (which is bounded to [0, 1)) so that it can appear anywhere in the image. This is what makes the sampling random. Finally, we cast the start coordinate as an int data type since multiplying by np.random.rand will return a float. Since we're working with images, we need an integer that corresponds to pixels. The end coordinate is just the start coordinate summed with the new image size. I suggest taking a look at PIL's crop function to see how this all works. We use the crop function along with the start and stop coordinates to randomly sample the image and its corresponding mask.  Lastly, this function also returns a mean value of the sampled image. The original images are incomplete for around 100 images. They have large white spaces but the masks still have ground truths in those areas. This is no good as it will only skew our parameters when we're training the U-Net. A simple way to get around this is to verify the mean of the whole image. Since large patches of white would increase the mean, we can set a threshold value to ensure that anything above that threshold will not be accepted as a sampled image (since it most likely contains a large area of white caused by the occlusions in the original image).

Following this function, it's just a basic for loop. The for loop is set to iterate based on the number of samples, in this case, 30,000. The for loop value is what we're going to use as the name so we set our range to be (1, number of samples + 1). We initialize the threshold value. Since the get_randompatch() function returns a for loop, this will get overwritten. To be honest, that should be initialized OUTSIDE the loop, I will fix that in future commits. You can see that I set a threshold value of 150. I chose this value after a couple of test runs, it seemed to provide the best trade off in terms of filtering out those large white areas I mentioned before. It will continuously sample until the threshold condition is met. Then it saves both the sampled image and mask to your sampled image & mask folders. The for loop iterates so you have 30k samples.

As long as your directories are set properly, all you would need to do is run this code and sit back until it's done

Hope this helped!

**Comment by Renjie8823** [comment]
emmm, actually I am very new to the deep learning, I know very little about it. I think it is because of the limitation of the internet, I cannot use your file to download the dataset. I can only use what I downloaded before. And as you know, the number of images and masks is different, so I deleted other redundant images, reduced them to 256*256(actually now I think it is wrong), and put them into the model. The result is very unclear. And now I did not know what to do next. If you have time , could you help me modify my code? I think it goes wrong in the data.py. Here is my code https://github.com/Renjie8823/deep-residual-unet.git  .Thanks very much.

**Comment by edwinpalegre** [comment]
Yeah that approach won't work. To comment on the mask and images, the number cannot be different. Each image has a corresponding mask, you MUST have the same number of images and masks. Resizing the original images to 256x256 also won't work well. You've made it so that when you feed your images into your network, the network won't properly pick up on the features since the image is so small. This is why a sampling method is a viable alternative approach. Unfortunately, I cannot commit to looking over your code as I have other commitments on my end. I would suggest a different project for now, since you don't have an extensive background on deep learning as of yet. I would also suggest writing the code from scratch and understanding the flow of the program.

I'll be closing this issue!
